torch_geometric/nn/conv/updated_gcn.py

```diff
@@ -0,0 +1,253 @@
+
+    def forward(self, x: Tensor, edge_index: Adj,
+                edge_weight: OptTensor = None, message_scale=None,
+                message_replacement=None) -> Tensor:
+        """"""
+
+        if self.normalize:
+            if isinstance(edge_index, Tensor):
+                cache = self._cached_edge_index
+                if cache is None:
+                    edge_index, edge_weight = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_edge_index = (edge_index, edge_weight)
+                else:
+                    edge_index, edge_weight = cache[0], cache[1]
+
+            elif isinstance(edge_index, SparseTensor):
+                cache = self._cached_adj_t
+                if cache is None:
+                    edge_index = gcn_norm(  # yapf: disable
+                        edge_index, edge_weight, x.size(self.node_dim),
+                        self.improved, self.add_self_loops)
+                    if self.cached:
+                        self._cached_adj_t = edge_index
+                else:
+                    edge_index = cache
+
+        if not self.explain:
+            x = self.lin(x)
+
+        # propagate_type: (x: Tensor, edge_weight: OptTensor)
+        out = self.propagate(edge_index, x=x, edge_weight=edge_weight,
+                             message_scale=message_scale,
+                             message_replacement=message_replacement,
+                             size=None)
+
+        if self.bias is not None and not False and not self.explain:
+            out += self.bias
+
+        return out
+
+    def message(self, x_i: Tensor, x_j: Tensor, edge_index: Tensor,
+                edge_weight: OptTensor, message_scale,
+                message_replacement) -> Tensor:
+        if self.explain:
+            basis_messages = x_j if edge_weight is None else edge_weight.view(
+                -1, 1) * x_j
+            basis_messages = self.convert(basis_messages)
+            basis_messages = self.process_message(basis_messages)
+
+            if message_scale is not None:
+                basis_messages = basis_messages * message_scale.unsqueeze(-1)
+
+                if message_replacement is not None:
+                    if basis_messages.shape == message_replacement.shape:
+                        basis_messages = basis_messages + \
+                                         (1 - message_scale).unsqueeze(
+                                             -1) * message_replacement
+                    else:
+                        basis_messages = basis_messages + \
+                                         ((1 - message_scale).unsqueeze(-1) *
+                                          message_replacement.unsqueeze(0))
+
+            self.latest_messages = basis_messages
+            self.latest_source_embeddings = x_j
```

Comment:
This is a bit dangerous to do. We should at least `detach()` them from the computation graph, right?

Reply:
These `self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings` are accessed through their respective functions `get_latest_messages()`, `get_latest_source_embeddings()`, and `get_latest_target_embeddings()` and used in Lines **_512-527_** of `GraphMask.py` for computing `edge_weights` to drop irrelevant edges using MLPs (after training of GNN Model). So, is it safe to `detach()` them from the computation graph?? Would love to know your thoughts on this.......

Reply:
Yeah, I agree. `detach()` would destroy the purpose of the `GraphMask` model. I think we have two options here:
* Save the features only in case `_explain=True`, and set them to `None` after optimization is done. Otherwise, the computation graph is never freed.
* A cleaner option is to never save them at all IMO. We should be able to make use of [message passing hooks](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.message_passing.MessagePassing.register_propagate_forward_pre_hook) inside `GraphMask` to obtain these features.

Reply:
Yes, we should go with the second point by not saving these features (`basis_messages`, `x_j`, and `x_i`) whatsoever. Can you please elaborate as to how we can obtain these features for each layer of the GNN model using message passing hooks?? Sorry, am bit novice to these hooks and don't know how to achieve this functionality.

Reply:
```python
def hook(module, inputs):
    print(inputs[-1].keys())

conv.register_message_forward_pre_hook(hook)
```
should work. Let me know about it!

Reply:
For GCN, this only prints (`x_j` and `edge_weight`) which is even expected though as `message()` of GCN has only these two parameters in its definition. Am still using the strategy, mentioned above, of saving these three features (`basis_messages`, `x_j`, and `x_i`) to three instance variables (`self.latest_messages`, `self.latest_source_embeddings`, and `self.latest_target_embeddings`). This simply saves all three features to the layer-specific instance variables and then we can access each of them for each layer of the GNN model while generating explanations. WDYT?? Yet, am still supporting and even looking forward to having such approach of accessing these three features for each layer using message passing hooks (and not saving three features at all) as the above approach:-

```
def hook(module, inputs):
    print(inputs[-1].keys())

conv.register_message_forward_pre_hook(hook)
```
appears to be not enough to consider as our solution. Let me know if we still have another way of achieving this or may be there is some possibility of tweaking this above approach to reach our solution of accessing those three features for each layer while not saving them in any instance variables. Would love to know your thoughts on this.......

Reply:
Indeed, let's go with `explain_message` :)